pallets/interest-accrual/src/lib.rs

```diff
@@ -0,0 +1,240 @@
+	Decrease(Amount),
+}
+
+// Type aliases
+type RateDetailsOf<T> = RateDetails<<T as Config>::InterestRate, Moment>;
+
+// Storage types
+#[derive(Encode, Decode, Default, Clone, PartialEq, TypeInfo)]
+#[cfg_attr(feature = "std", derive(Debug))]
+pub struct RateDetails<InterestRate, Moment> {
+	// chi in MCD Rates
+	pub cumulative_rate: InterestRate,
+
+	// when cumulative_rate was last updated
+	pub last_updated: Moment,
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub (super) trait Store)]
+	#[pallet::without_storage_info]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		type Balance: Member
+			+ Parameter
+			+ AtLeast32BitUnsigned
+			+ Default
+			+ Copy
+			+ MaxEncodedLen
+			+ FixedPointOperand
+			+ From<u64>
+			+ From<u128>
+			+ TypeInfo
+			+ TryInto<u64>;
+
+		/// A fixed-point number which represents
+		/// an interest rate.
+		type InterestRate: Member
+			+ Parameter
+			+ Default
+			+ Copy
+			+ TypeInfo
+			+ FixedPointNumber<Inner = Self::Balance>;
+
+		/// A fixed-point number which represents
+		/// the normalized debt.
+		type NormalizedDebt: Member
+			+ Parameter
+			+ Default
+			+ Copy
+			+ TypeInfo
+			+ FixedPointNumber
+			+ CheckedMul;
+
+		/// The amount type
+		type Amount: Parameter
+			+ Member
+			+ MaybeSerializeDeserialize
+			+ FixedPointNumber
+			+ FixedPointOperand
+			+ From<Self::NormalizedDebt>
+			+ CheckedAdd
+			+ CheckedSub
+			+ TypeInfo;
+
+		type Time: UnixTime;
+	}
+
+	#[pallet::storage]
+	#[pallet::getter(fn get_sale)]
+	pub(super) type Rates<T: Config> =
+		StorageMap<_, Blake2_128Concat, T::InterestRate, RateDetailsOf<T>, OptionQuery>;
+
+	#[pallet::event]
+	pub enum Event<T: Config> {}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Emits when the debt calculation failed
+		DebtCalculationFailed,
+
+		/// Emits when the debt adjustment failed
+		DebtAdjustmentFailed,
+
+		/// Emits when the interest rate was not used
+		NoSuchRate,
+	}
+
+	// TODO: add permissionless extrinsic to update any rate
+
+	impl<T: Config> Pallet<T> {
+		pub fn do_get_current_debt(
+			interest_rate_per_sec: T::InterestRate,
+			normalized_debt: T::NormalizedDebt,
+		) -> Result<T::Amount, DispatchError> {
+			let rate = match Rates::<T>::try_get(interest_rate_per_sec) {
+				Err(_) => {
+					let new_rate = RateDetails {
+						cumulative_rate: T::InterestRate::saturating_from_rational(100, 100).into(),
+						last_updated: Self::now(),
+					};
+					Rates::<T>::insert(interest_rate_per_sec, &new_rate);
+					new_rate.cumulative_rate
+				}
+				Ok(rate) => {
+					let new_cumulative_rate = Self::calculate_cumulative_rate(
+						interest_rate_per_sec,
+						rate.cumulative_rate,
+						rate.last_updated,
+					)
+					.ok_or(Error::<T>::DebtCalculationFailed)?;
```

Comment:
I am missing an update of ther underlying `RateDetails` here, or in the `calculate_cumulative_rate`. I would prefer the later.